### ISSUE TITLE
LDAC Choppiness/stuttering 

### ISSUE THREAD

**Issue by trip6kitty** [Pittvandewitt/Wavelet#244]
I noticed that when I use my Sony WFXM4 ear buds that have LDAC, while listening with LDAC the audio will come very choppy and stutter.

I know turning off LDAC resolves it, but of course I want to listen with LDAC on.

I recently tried turning off Wavelet when experiencing this issue and that also fixed it (but of course I want to listen with wavelet).

I noticed that I experience the issue more when the phone is locked and not in use. The issue sometimes goes away when I unlock the phone and use it. 

Are you aware of this issue?

I have a Pixel 6 pro on Android 14 beta 1, but I have experienced this issue for a while when also on Android 13.

Thank you.  Please let me know if you need anything from me and I will be able to assist.

**Comment by Azel04** [comment]
I can confirm this behaviour too.
I have a Pixel 6 on Android 14 (Stable) and I am using the FiiO BTR5 2021

**Comment by Pittvandewitt** [comment]
I've tried replicating the issue but without success. WH-1000XM4 headphones and Android 14 on a Pixel 7 and Android 12 on an LG V50. Which settings do you have enabled, and to what value is the buffer size set to?

**Comment by Azel04** [comment]
I just have AutoEQ enabled (Moondrop Aria Snow Edition profile) with 4096 samples.

I'll try to replicate the issue again (perhaps it got fixed with the Android 14 November Update?)

**Comment by trip6kitty** [comment]
Buffer size was default.  AutoEQ enabled for WF XM10004 and Graphic
Equalizer set to V shape.

I have since got a new phone and the issue is gone (I now have a P8 Pro).

I had found a cause and workaround though, in my anecdotal experience.  If
I started listening to music on one app (e.g. Spotify), then switched to
another (e.g. Poweramp), I would experience the issue when the screen is
off/locked.  To fix it, I would have to kill the first app (Spotify).  So I
unlock, swipe down notifications completely, then at the bottom where it
says  "2 apps are active", I would open that then kill the first app
(Spotify).  I have found it doesn't matter what the app is, I have
experienced it with Spotify, Poweramp, Soundcloud, and Youtube.  The
resolution is the same: kill the app that I am no longer listening to.

**Comment by Azel04** [comment]
I can confirm what you said here.

Having 2 Apps open (in my case Spotify and Synfonium) and then turning the screen off results in choppiness/stuttering with LDAC.

Killing one of the apps fixes the issue.

**Comment by trip6kitty** [comment]
Strangely it only happens on P6 Pro. This does not happen on my Pixel 8
Pro. Both are on the same version of Android.

**Comment by Azel04** [comment]
I have a Pixel 6... Weird

**Comment by hmnd** [comment]
I'm experiencing the same on Android 14 on a Pixel 7 with XM3's. Can confirm @trip6kitty trick of killing the app no longer in use fixes it :thinking:

**Comment by ElfyTR** [comment]
I have the same issue with the wh-1000xm5 and galaxy buds 2 pro even wrote a issue too has anyone been able to solve this

**Comment by jhodgdev** [comment]
Experiencing the same issue on Wavelet 23.09

- Google Pixel 6 Pro
- Sony WH-1000XM5 w/ incl. AutoEq profile.
- Occurs with buffer size at 4096 and 8192 samples.
- Choppiness only occurs when phone is locked.

Killing apps so only one is using Wavelet --- in my case, YouTube and Pocket Casts --- does resolve the problem.

**Comment by NielsMayer** [comment]
Instead of killing other apps, what about turning on "power saving exclusions" for wavelet and any media players that might get put into "power saving mode" especially beginning with Android 13 and 14...

![Screenshot_20240330-110308](https://github.com/Pittvandewitt/Wavelet/assets/6306083/fdbe8632-d5d9-4b75-8dac-993b84426078)

......

some notes I wrote on this topic:

https://hyperionics.com/atVoice/stops.html

Note @voice launches the "power savings exclusions" settings directly on clicking "continue" in the second dialog after clicking "continue" in the first. 

https://dontkillmyapp.com/
--  samsung especially problematic
https://dontkillmyapp.com/samsung
https://dontkillmyapp.com/general

Aosp is best
https://dontkillmyapp.com/google

"Android P
There’s a special option in Settings > Apps > Your app > Advanced > Battery > Background restrictions. If users accidentally enable this option, it will break their apps. And users do enable that option!

Pie and pre-Pie
If you see background processing issues, overall it is a good idea to make your app not battery optimized to ensure it gets the freedom it needs to perform in the background.

For that:

Go to Settings > Apps > Your app > Advanced > Battery > Battery optimization

Change view to All apps

Search for your app

Choose Not optimized

Android O
Make sure Settings > Apps > Your app > Advanced > Battery > Background limitations is not enabled. If the app is not yet optimized for Oreo API level it will break their background processing."

**Comment by trip6kitty** [comment]
Revisiting this as I recently got WH XM5s and was running into this issue again.  The "workaround" I commented on 11/14/23 still works, however the comment by @NielsMayer reminded me to check the battery restriction of the Wavelet app.  I saw that it was set to Restricted.  Once I set to Unrestricted, it appears to have fixed the issue.

**Comment by NielsMayer** [comment]
@trip6kitty -- This issue really shouldn't be closed. Has a mechanism been added to check the status of battery-saving and pop a dialog warning the use, similar to what the "@voice" app does? 

My comments (pasted from my existing notes on the issue) and the arrived-at workaround were posted here because I am also facing the need to fix this issue in my own app. 

I'd also like to get opinions from the developer @Pittvandewitt  and other users on the correct way to go about fixing this issue in an app that might target a wide range of devices and types of devices. For example Android TV seems to handle this kind of thing differently, given that "TV" typically runs with line power. However, if the app is intended to run in the background, that could impose other similar issues. 

Discuss. :-)

**Comment by trip6kitty** [comment]
Yes I didn't realize that I would close the issue upon commenting. 

Also I saw that my issue was not resolved either, as even with Unrestricted battery usage on Wavelet the issue came back. 

I'll reopen the issue.

**Comment by trip6kitty** [comment]
Wondering if anyone has a resolution?  I have this issue still with the latest Wavelet version and Sony WH XM5s (as well as WF XM4).

**Comment by Shinzef** [comment]
Can confirm this happens in LDAC enabled earphones in general. Whenever wavelet is on, it introduces skips and crackling. Some earphones have worse degradation and others fair better.

**Comment by qumusabel** [comment]
Same issue appeared on my Pixel 8 Pro after updating to Android 17. Had no such problems on A16 with my WH-1000XM4s. Apple Music is problematic in particular. Sometimes it seems to work okay until I open another app that might want to play sound (like Chrome). Switching to AAC helps but that setting does not stick on Pixels.

**Comment by trip6kitty** [comment]
Yeah I noticed that the issue returned after updating to Android 17. Prior to updating to A17, it did not happen for a very long time.

**Comment by mael-queau** [comment]
Hi, I can confirm that this also happens on my Pixel 10 Pro XL using Sony WH-1000XM4 headphones, on both stock OS and GrapheneOS. I've been having issues with the Pixel series and LDAC for a few generations, but this time around the stuttering completely goes away anytime I switch off AutoEQ in Wavelet or switch to AAC for instance, which indicated to me that there might be an issue within Wavelet in addition to the usual Pixel-related issues, at least in my situation
